review pass on tutorial sections (#806)
NixOS/nix.dev: source/tutorials/file-sets.md

```
:::

Even though file sets conceptually contain local files, these files are *never* added to the Nix store unless explicitly requested.
You don't have to worry about accidentally copying secrets into the world-readable store.

In this example, although we pretty-printed the home directory, no files were copied.
This is also evident from the expression evaluating instantly.

:::{note}
This is in contrast to coercion of paths to strings such as in `"${./.}"`,
which copies the whole directory to the Nix store on evaluation!
:::

:::{warning}

> With the [`flakes` experimental feature](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake) enabled, a local directory containing `flake.nix` is always copied into the Nix store *completely* unless it is a Git repository!

Author: Oof. But what are "experimental Flakes"?

Unrelated, I think `flakes` should transitively enable `nix-command`, @Ericson2314 wdyt?

> With the [`flakes` experimental feature](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake) enabled, a local directory containing `flake.nix` is always copied into the Nix store *completely* unless it is a Git repository!

Reviewer: Alternative wordings welcome, I think "experimental Flakes" with the link should be clear enough for now though. I'd rather not detangle Flakes in this tutorial :P

trace: - nix.conf (symlink)
```

In addition to the included file, this also prints its [file type](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-readFileType).

If a given path doesn't exist, the library will complain:

```shell-session
nix-repl> fs.trace /etc/nix/nix.nix
error: lib.fileset.trace: Argument (/etc/nix/nix.nix)
  is a path that does not exist.
```

## Example project

To further experiment with the library, make a sample project.
Create a new directory, enter it,
and set up `niv` to pin the Nixpkgs dependency:

```shell-session
$ mkdir fileset
$ cd fileset
$ nix-shell -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch nixos-unstable"
```

:::{note}
We're using the `nixos-unstable` channel branch here, since no stable release has all the features needed for this tutorial.
:::

Then create a `default.nix` file with the following contents:

```{code-block} nix
:caption: default.nix
{
  system ? builtins.currentSystem,

>   inputs ? import ./nix/sources.nix,

Author: Ah, oh. Apparently I refrained from actually committing to that, but it stuck after spending a bit of time thinking it through. Now I remember: I didn't pull through because it would diverge from `niv` upstream docs and likely confuse people. Sorry for the detour.

(I still think it should consider converging with flake conventions, `inputs` or `sources`, either way. `sources` has some appeal because this is what the references are about: remote sources of... stuff. `inputs` is so generic to the point it's meaningless; and they aren't even necessarily used as build inputs.)

  To create a file set from a path that may not exist, use `lib.fileset.maybeMissing`.
```

Follow the instructions in the error message, and use `maybeMissing` to create a file set from a path that may not exist (in which case the file set will be empty):

<!-- https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.maybeMissing -->

```{code-block} diff
:caption: build.nix
 { stdenv, lib }:
 let
   fs = lib.fileset;
-  sourceFiles = fs.difference ./. ./result;
+  sourceFiles = fs.difference ./. (fs.maybeMissing ./result);
 in
```

This now works, using the whole directory since `./result` is not present:

```
$ nix-build
trace: /home/user/select (all files in directory)
this derivation will be built:
  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
...
/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
```

Another build attempt will produce a different trace, but the same output path:

```
$ nix-build
trace: /home/user/fileset
trace: - build.nix (regular)
trace: - default.nix (regular)
trace: - hello.txt (regular)
trace: - nix (all files in directory)
trace: - world.txt (regular)
/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
```

> ## Union (exclude)

Reviewer: ```suggestion
## Union (for excluding only select files)
```

Though this also makes it sound like that's what this function does, which would be wrong :thinking:

Author: ```suggestion
## Union (explicitly exclude files)
```

?

Reviewer: I like it!

+        ./nix
+      ]);
 in
```

Changing any of the excluded files now doesn't necessarily require a rebuild anymore.

Check it and modify one of the excluded files again:

```
$ echo >> build.nix
```

```
$ nix-build
trace: /home/user/fileset
trace: - hello.txt (regular)
trace: - world.txt (regular)
/nix/store/ckn40y7hgqphhbhyrq64h9r6rvdh973r-fileset
```

## Filter

Dealing with a large number of files, independent of their location, can be done programmatically with the [`fileFilter`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.fileFilter) function.

Use it to select all files with a name ending in `.nix`:

```{code-block} diff
:caption: build.nix
 { stdenv, lib }:
 let
   fs = lib.fileset;
   sourceFiles =
     fs.difference
       ./.
       (fs.unions [
         (fs.maybeMissing ./result)
-        ./default.nix
-        ./build.nix
+        (fs.fileFilter (file: lib.hasSuffix ".nix" file.name) ./.)
         ./nix
       ]);
 in
```

This does not change the result:

```shell-session
$ nix-build
trace: /home/user/fileset
trace: - hello.txt (regular)
trace: - world.txt (regular)
/nix/store/ckn40y7hgqphhbhyrq64h9r6rvdh973r-fileset
```

> Notably, with this approach, new files added to the source directory are _included_ by default.
> Depending on your project, this might be a better fit than the alternative in the next section.

Reviewer: ```suggestion
Notably, with this approach of using `difference ./.`, we're explicitly selecting the files to _exclude_, which means that new files added to the source directory are _included_ by default.
Depending on your project, this might be a better fit than the alternative in the next section.
```

Author: ```suggestion
Notably, the approach of using `difference ./.` explicitly selects the files to _exclude_, which means that new files added to the source directory are included by default.
Depending on your project, this might be a better fit than the alternative in the next section.
```

Removing the ominous "we" and the emphasis on "including", which was a bit irritating already because the section is "Union (exclude)" - now that's resolved, yay!

> ## Union (include)

Reviewer: ```suggestion
## Union (for including only select files)
```

To avoid confusion with the previous paragraph mentioning "included by default"

Author: That would be resolved with the other suggestion.

Reviewer: ```suggestion
## Union (explicitly include files)
```

Let's do this though, I already committed the `explicitly exclude files` now

      ./src
    )
  ];
in

fs.trace sourceFiles

stdenv.mkDerivation {
  name = "fileset";
  src = fs.toSource {
    root = ./.;
    fileset = sourceFiles;
  };
  postInstall = ''

>     cp -vr $src $out

Reviewer: ```suggestion
    cp -vr . $out
```

The source will be in the local directory

Author: Yeah but that's another layer of indirection. With `$src` you "only" have to know that non-reserved attrs become env vars to fully grasp the whole thing. Now you'd also have to know that other bit.

Reviewer: We're already using that fact throughout the entire tutorial though, starting with the first `cp -v hello.txt $out`. It should be consistent, but I also don't want those long `/nix/store/... ->` lines in the log output.

  '';
}
```

The `postInstall` script is simplified to rely on the sources to be pre-filtered appropriately:

```shell-session
$ nix-build
trace: /home/user/fileset
trace: - build.sh (regular)
trace: - hello.txt (regular)
trace: - src (all files in directory)
trace: - world.txt (regular)
this derivation will be built:
  /nix/store/sjzkn07d6a4qfp60p6dc64pzvmmdafff-fileset.drv
...

> '/nix/store/6k6pv78b1dkdhbdlzxar5vb3z3c8fwza-source' -> '/nix/store/zl4n1g6is4cmsqf02
> dci5b2h5zd0ia4r-fileset'
> '/nix/store/6k6pv78b1dkdhbdlzxar5vb3z3c8fwza-source/build.sh' -> '/nix/store/zl4n1g6i
> s4cmsqf02dci5b2h5zd0ia4r-fileset/build.sh'
> '/nix/store/6k6pv78b1dkdhbdlzxar5vb3z3c8fwza-source/hello.txt' -> '/nix/store/zl4n1g6
> is4cmsqf02dci5b2h5zd0ia4r-fileset/hello.txt'
> '/nix/store/6k6pv78b1dkdhbdlzxar5vb3z3c8fwza-source/world.txt' -> '/nix/store/zl4n1g6
> is4cmsqf02dci5b2h5zd0ia4r-fileset/world.txt'
> '/nix/store/6k6pv78b1dkdhbdlzxar5vb3z3c8fwza-source/src' -> '/nix/store/zl4n1g6is4cms
> qf02dci5b2h5zd0ia4r-fileset/src'
> '/nix/store/6k6pv78b1dkdhbdlzxar5vb3z3c8fwza-source/src/select.c' -> '/nix/store/zl4n
> 1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src/select.c'
> '/nix/store/6k6pv78b1dkdhbdlzxar5vb3z3c8fwza-source/src/select.h' -> '/nix/store/zl4n
> 1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src/select.h'

Reviewer: ```suggestion
'.' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset'
'./build.sh' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/build.sh'
'./hello.txt' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/hello.txt'
'./world.txt' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/world.txt'
'./src' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src'
'./src/select.c' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src/select.c'
'./src/select.h' -> '/nix/store/zl4n1g6is4cmsqf02dci5b2h5zd0ia4r-fileset/src/select.h'
```

With the `cp` change

Author: Eh. Okay, that's actually a good argument to do that.

```shell-session
$ nix-build
warning: Git tree '/home/user/select' is dirty
trace: /home/vg/src/nix.dev/fileset
trace: - build.nix (regular)
trace: - build.sh (regular)
trace: - default.nix (regular)
trace: - hello.txt (regular)
trace: - nix (all files in directory)
trace: - src
trace:   - select.c (regular)
trace:   - select.h (regular)
trace: - world.txt (regular)
this derivation will be built:
  /nix/store/p9aw3fl5xcjbgg9yagykywvskzgrmk5y-fileset.drv
...
/nix/store/cw4bza1r27iimzrdbfl4yn5xr36d6k5l-fileset
```

This includes too much though, as not all of these files are needed to build the derivation as originally intended.

:::{note}

> With the [`flakes` experimental feature](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake) enabled, it's [not really possible](https://github.com/NixOS/nix/issues/9292) to use this function, even with

Reviewer: ```suggestion
With [experimental Flakes](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake), it's [not really possible](https://github.com/NixOS/nix/issues/9292) to use this function, even with
```
